### PR TITLE
Add Event Hub E2E test

### PR DESCRIPTION
**Pull request by Flanker32**
Add test case for eventhub: #261

### CODE REVIEW

File: azure-functions-maven-plugin/pom.xml [hunk 1]
```diff
@@ -236,6 +236,16 @@
             <version>${spring-test.version}</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>com.microsoft.azure</groupId>
```
Comment (jdneo) [inline azure-functions-maven-plugin/pom.xml]: set `<scope>` to `test`

File: azure-functions-maven-plugin/pom.xml [hunk 2]
```diff
@@ -236,6 +236,16 @@
             <version>${spring-test.version}</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>com.microsoft.azure</groupId>
+            <artifactId>azure-eventhubs</artifactId>
+            <version>1.2.0</version>
+        </dependency>
+        <dependency>
+            <groupId>com.microsoft.azure</groupId>
```
Comment (jdneo) [inline azure-functions-maven-plugin/pom.xml]: set `<scope>` to `test`

File: azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/invoker/storage/EventProcessor.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for
+ * license information.
+ */
+
+package com.microsoft.azure.maven.function.invoker.storage;
```
Comment (jdneo) [inline azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/invoker/storage/EventProcessor.java]: You can create a new package `com.microsoft.azure.maven.function.invoker.eventhub` and put this file into it.

File: azure-functions-maven-plugin/src/it/3-eventhub-trigger/cleanup.groovy [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+String resourceGroup = "maven-functions-it-${timestamp}-rg-3"
+
+EventHubProcesser eventHubProcesser = null
+try{
+    eventHubProcesser = new EventHubProcesser(resourceGroup,nameSpace,storageName);
+    eventHubProcesser.createOrGetEventHubByName("trigger")
+    eventHubProcesser.createOrGetEventHubByName("output")
+    // Get connnection string of EventHub and set it to trigger function
+    def connectionString = eventHubProcesser.getEventHubConnectionString()
+    CommonUtils.executeCommand("az webapp config appsettings set --name ${functionName} --resource-group ${resourceGroup} --settings CIEventHubConnection=\"${connectionString}\"")
```
Comment (SummerSun) [inline azure-functions-maven-plugin/src/it/3-eventhub-trigger/cleanup.groovy]: Check does it need the `az login` first.

File: azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/invoker/storage/EventHubProcesser.java [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    public List<String> getMessageFromEventHub(String eventhub) throws Exception {
+        final List<String> result = new ArrayList<>();
+        final EventHubClient ehClient = getEventHubClientByName(eventhub);
+        final List<String> partitionIds = Arrays.asList(ehClient.getRuntimeInformation().get().getPartitionIds());
+        partitionIds.parallelStream()
+                .forEach(partitionId -> result.addAll(receiveMessageFromSpecificPartition(ehClient, partitionId)));
+        return result;
+    }
+
+    public List<String> receiveMessageFromSpecificPartition(EventHubClient eventHubClient, String partitionId) {
```
Comment (SummerSun) [inline azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/invoker/storage/EventHubProcesser.java]: `receiveMessageFromPartition` is ok

File: azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/invoker/storage/EventHubProcesser.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    public void sendMessageToEventHub(String eventhub, String message) throws Exception {
+        final EventHubClient ehClient = getEventHubClientByName(eventhub);
+        final String payload = message;
+        final Gson gson = new Gson();
+        final byte[] payloadBytes = gson.toJson(payload).getBytes(Charset.defaultCharset());
+        final EventData sendEvent = EventData.create(payloadBytes);
+        ehClient.send(sendEvent).get();
+    }
+
+    public List<String> getMessageFromEventHub(String eventhub) throws Exception {
```
Comment (SummerSun) [inline azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/invoker/storage/EventHubProcesser.java]: Align the `getMessage` or `receiveMessage` with the following methods

File: azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/invoker/storage/EventHubProcesser.java (hunk 5 above)
Comment (SummerSun) [inline azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/invoker/storage/EventHubProcesser.java]: Add `@NotNull` to the parameters to leverage the IDE NPE check. And better to add the `final` modifier, even though these are test code, it is better if we apply the same style with production code.

File: azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/invoker/storage/EventHubProcesser.java [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+        if (resultList.size() == 0) {
+            return azureClient.eventHubs().define(eventHubName).withExistingNamespace(eventHubNamespace)
+                    .withExistingStorageAccountForCapturedData(storageAccount, eventHubName)
+                    .withDataCaptureEnabled().create();
+        } else {
+            return resultList.get(0);
+        }
+    }
+
+    public void sendMessageToEventHub(String eventhub, String message) throws Exception {
```
Comment (SummerSun) [inline azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/invoker/storage/EventHubProcesser.java]: Parameter `eventHubName`.

File: azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/invoker/storage/EventHubProcesser.java [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+        final EventHubClient ehClient = getEventHubClientByName(eventhub);
```
Comment (SummerSun) [inline azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/invoker/storage/EventHubProcesser.java]: Suggest `eventHubClient`

File: azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/invoker/storage/EventHubProcesser.java [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+        final String payload = message;
```
Comment (SummerSun) [inline azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/invoker/storage/EventHubProcesser.java]: What's the point?

Comment (Flanker32) [reply azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/invoker/storage/EventHubProcesser.java]: Follow by Azure Document: https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-java-get-started-send#send-events

File: azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/invoker/storage/EventHubProcesser.java [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+
+    public void close() throws Exception {
+        for (final EventHubClient eventHubClient : eventHubClientMap.values()) {
+            eventHubClient.closeSync();
+        }
+        executorService.shutdown();
+    }
+
+    private EventHubClient getEventHubClientByName(String eventHubName) throws Exception {
+        if (!eventHubClientMap.containsKey(eventHubName)) {
```
Comment (SummerSun) [inline azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/invoker/storage/EventHubProcesser.java]: Suggest the other way around:
```Java
if (eventHubClientMap.containsKey(eventHubName)) {
    return eventHubClientMap.get(eventHubName);
}
// handle the key not exists case
```

File: azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/invoker/storage/EventHubProcesser.java [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+    private EventHubNamespace eventHubNamespace;
+    private StorageAccount storageAccount;
+    private ResourceGroup resourceGroup;
+
+    private ExecutorService executorService = Executors.newCachedThreadPool();
+    private Map<String, EventHubClient> eventHubClientMap = new HashMap<>();
+
+    public EventHubProcesser(String resourceGroupName, String namespaceName, String storageAccountName)
+            throws Exception {
+        final Azure azureClint = CommonUtils.getAzureClient();
```
Comment (SummerSun) [inline azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/invoker/storage/EventHubProcesser.java]: Typo 'azureClient`

File: azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/invoker/storage/EventHubProcesser.java [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+
+        if (azureClint.resourceGroups().contain(resourceGroupName)) {
+            resourceGroup = azureClint.resourceGroups().getByName(resourceGroupName);
+        } else {
+            resourceGroup = azureClint.resourceGroups().define(resourceGroupName).withRegion(Region.US_EAST).create();
+        }
+
+        final boolean isEventHubNamespaceExist = azureClint.eventHubNamespaces().list().stream()
```
Comment (SummerSun) [inline azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/invoker/storage/EventHubProcesser.java]: Why not use the same way as getting the `storageAccount`? Try to get the value first, if the return value is null, then to create.

File: azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/invoker/storage/EventHubProcesser.java [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+                    .withSku(StorageAccountSkuType.STANDARD_LRS)
+                    .withGeneralPurposeAccountKindV2().create();
+        }
+    }
+
+    public EventHub createOrGetEventHubByName(String eventHubName) throws Exception {
+        final Azure azureClient = CommonUtils.getAzureClient();
+        final List<EventHub> resultList = eventHubNamespace.listEventHubs().stream()
+                .filter(eventHub -> eventHub.name().equals(eventHubName)).collect(Collectors.toList());
+        if (resultList.size() == 0) {
```
Comment (SummerSun) [inline azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/invoker/storage/EventHubProcesser.java]: Are you trying to get the first match? Use the `.filter().findFirst().get()`

File: azure-functions-maven-plugin/src/it/3-eventhub-trigger/cleanup.groovy [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+ * license information.
+ */
+
+import com.microsoft.azure.maven.function.invoker.CommonUtils
+import com.microsoft.azure.maven.function.invoker.storage.EventHubProcesser
+
+String functionName = "maven-functions-it-${timestamp}-3"
+String storageName = "cihub${timestamp}"
+String nameSpace = "FunctionCIEventHubNamespace-${timestamp}"
+String resourceGroup = "maven-functions-it-${timestamp}-rg-3"
```
Comment (SummerSun) [inline azure-functions-maven-plugin/src/it/3-eventhub-trigger/cleanup.groovy]: resourceGroupName

File: azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/invoker/storage/EventHubProcesser.java (hunk 10 above)
Comment (Flanker32) [reply azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/invoker/storage/EventHubProcesser.java]: I prefer the origin design for i just do single return in this method, while i will write two return statement in this method if i process contains first.

File: azure-functions-maven-plugin/src/it/3-eventhub-trigger/cleanup.groovy [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+
+EventHubProcesser eventHubProcesser = null
+try{
+    eventHubProcesser = new EventHubProcesser(resourceGroup,nameSpace,storageName);
+    eventHubProcesser.createOrGetEventHubByName("trigger")
```
[the remark this answers is not in the document]
Comment (Flanker32) [reply azure-functions-maven-plugin/src/it/3-eventhub-trigger/cleanup.groovy]: i do need to create the eventhub in verify part and the function means get the eventhub, if not exist , create it and return.

**Pull request merged**